### ISSUE TITLE
Setting up a branching system for the project

### ISSUE THREAD

**Issue by SunyataZero** [mindfulness-at-the-computer/mindfulness-at-the-computer#97]
* [x] deciding on what approach we want
* [x] Also updating the documentation

**Comment by JedMeister** [comment]
I quite like some of the more complex branching methodologies that are around. Although sometimes, keeping it simple is best way to go!?!

Regardless, the best branching technique for this project will really depend on your release model. My suggestion would be to only ever have one "supported" release (i.e. current stable; although you can do beta/rc releases in the same way). If you do that, you will never need to update old code to include new bugfixes (everybody needs to update to the latest version to get the fixes).

It's probably not ideal on large, complex professional software projects, but I think for something like this it's fine. 

If you go that path, then generally, there isn't really any need for any branches other than master. If you encourage all contributors to use an alternate branch (i.e. anything other than master) when they do their development, then even doing quite complex merges of multiple (remote) branches is possible. You could then argue that each contributor is working on their own development branch!

If you get to a point where a particular change is going to be pretty radical and/or will take quite a while and/or multiple people are working on it, you can create (organisational) development branches on an adhoc basis. Once they're merged, just delete them again...

You can use tags to denote specific releases, from minor bugfixes to major version bumps. They are your stable points. Ideally the HEAD of master should always build, but it's essentially your development branch. That way you don't need a "stable" or "production" branch, if people want stable, they can pull by tag!

My 2c anyway... :smile:

**Comment by SunyataZero** [comment]
@JedMeister Thank you for your help! I agree about the release model, only one release should be supported. (About the using a single branch: I think i agree but need to consider this some more, there's a lot of new data for me to process :) )

**Comment by rogermeadows** [comment]
@JedMeister Thanks for you input. I agree with supporting only one release. The main reason I am inclined to use multiple branches has to do with having a easy way to identify the level of stability (amount of successful testing) associated with a commit. However, there are other ways we could identify stability. We might use tags on the master branch to identify levels of stability. I am less concerned with the details of how we identify stability and more concerned with having a clearly defined and followed stability path. 

I suggest that automated testing (unit testing and/or GUI testing) is a necessary component of development work. In other words, a piece of new code would not be considered complete until it included new unit and/or GUI automated tests covering the new code. In addition, all previous unit and GUI tests must also pass before the new code is considered ready to be merged into the main repository. 

It seems that we have a start in that direction with the Travis CI connection that can automatically run tests for each push to a developers fork and for each pull request. 

As you can see, I am a big proponent of automated unit and when appropriate automated GUI testing.

My 2c ...

**Comment by JedMeister** [comment]
Automated testing FTW! There should be much more of it IMO! :smile: 

As I hinted (perhaps not strongly enough). I think that major changes _should_ be developed in a separate development branch! IMO only bugfixes and minor modifications should be merged straight into master (and only after they've passed any tests). As I said, ideally the HEAD of master should always build successfully!

I really like Vincent Driessen's [GitFlow](https://datasift.github.io/gitflow/IntroducingGitFlow.html) but IMO it's often overkill. In my experience, having a specific ongoing/permanent development branch doesn't really give any real advantage and has additional overhead costs. 

Assuming that most bugfixes and minor features will be developed by only one person at a time, then the "feature" and "hotfix" branches that Vincent refers to still exist. But instead of being specific branches within the organisational repo, they're the branches of the particular developer(s) code fork that they are working in. The dev then opens a PR (against org/master) and once it passes review and tests, then it's merged.

Once you get multiple people working on the one thing, that's when an organisational level branch becomes desirable IMO.

Bottom line for me is that premature optimisation is the root of all evil! :stuck_out_tongue: 

Don't get me wrong, I think clear processes are a good thing, but sometimes too much focus on process (especially if it's a rigid process) just creates additional overhead and more friction, especially for new contributors. 

Another 2c...! :smile:
